File: packages/sdk-ts/src/tools.ts
/**
 * The single source of the four tools.
 *
 * Four, not five: `execute_metered` disappeared along with the routine regime
 * (docs/09 § 2, docs/03 § "One regime, two rails"). Four, not fifteen: the
 * catalogue is deliberately small, because an agent chooses badly among fifteen
 * tools.
 *
 * Everything below is framework-agnostic. The MCP server and the Vercel AI
 * adapter do nothing but project these descriptors — that is the "one single
 * source, adapters under 100 lines" structure that won the previous hackathon
 * (docs/09 § intro).
 */

import { validateActionSpec } from '@warrant/core'
import type { ActionSpec, Address, Hex } from '@warrant/core'
import { z } from 'zod'

import { WarrantError, toWarrantError } from './errors.js'
import type {
  GatewayClient,
  ListWarrantsResult,
  QuoteResult,
  WarrantOpened,
  WarrantView,
} from './gateway.js'
import {
  getWarrantInputSchema,
  getWarrantOutputSchema,
  listWarrantsInputSchema,
  listWarrantsOutputSchema,
  quoteRiskInputSchema,
  quoteRiskOutputSchema,
  requestWarrantInputSchema,
  requestWarrantOutputSchema,
} from './schemas.js'
import type {
  PaymentPayload,
  PaymentRequired,
  PaymentSigner,
  SettlementResponse,
} from './x402.js'

export interface ToolContext {
  /** Payment supplied by the client — on MCP, `_meta["x402/payment"]`. */
  payment?: PaymentPayload | undefined
}

/**
 * Outcome of a tool call.
 *
 * `payment-required` is not an error from the domain's point of view: it is
 * step 2 of the x402 transport (docs/05 § 1.7). Adapters are free to represent
 * it as an error — MCP requires it — but the core does not.
 */
export type ToolOutcome<T> =
  | { kind: 'ok'; data: T; settlement?: SettlementResponse | undefined }
  | { kind: 'payment-required'; paymentRequired: PaymentRequired }

export interface WarrantTool<TInput extends z.ZodObject<z.ZodRawShape>, TOutput> {
  name: string
  title: string
  description: string
  input: TInput
  /** Minimal output contract, published as `outputSchema` by `tools/list`. */
  output: z.ZodObject<z.ZodRawShape>
  /** `true` for `request_warrant` alone: the bond must be funded. */
  paid: boolean
  /** None of these tools mutates onchain state beyond the payment itself. */
  readOnly: boolean
  run(client: GatewayClient, args: unknown, ctx?: ToolContext): Promise<ToolOutcome<TOutput>>
}

/**
 * Parses and **strips** the arguments.
 *
 * The stripping is the security point: the returned value contains only the
 * keys of the schema, so a `category` field slipped into the `actionSpec` can
 * reach neither the Classifier nor the `actionHash`.
 */
function parseArgs<S extends z.ZodObject<z.ZodRawShape>>(schema: S, args: unknown): z.output<S> {
  const parsed = schema.safeParse(args)
  if (parsed.success) return parsed.data

  const first = parsed.error.issues[0]
  const field = first ? `$.${first.path.join('.')}` : '$'
  const message = parsed.error.issues
    .map((i) => `$.${i.path.join('.')}: ${i.message}`)
    .join('; ')
  throw new WarrantError('invalid_input', message, {
    field,
    hint: `Fix ${field} then call the tool again. Reminder: neither category nor notional is accepted — both are derived from the calldata.`,
    details: parsed.error.issues,
  })
}

/**
 * Double validation of the `actionSpec`: Zod for the shape, `@warrant/core` for
 * the domain invariants.
 *
 * Redundant in appearance, useful in practice: the core is the authority on what
 * is hashable, and it is the core that names the offending field exactly as the
 * Gateway will. Letting the two validations diverge means accepting that a
 * warrant be accepted here and rejected three milliseconds later.
 */
function toActionSpec(spec: unknown): ActionSpec {
  try {
    return validateActionSpec(spec)
  } catch (err) {
    throw toWarrantError(err)
  }
}

async function callGateway<T>(fn: () => Promise<T>): Promise<T> {
  try {
    return await fn()
  } catch (err) {
    throw toWarrantError(err)
  }
}

// ─────────────────────────────────────────────────────────────────────────────

export const quoteRiskTool: WarrantTool<typeof quoteRiskInputSchema, QuoteResult> = {
  name: 'quote_risk',
  title: 'Quote the bond for an action',
  description:
    'Estimates the bond required for an action, committing nothing and paying nothing. Classifies the calldata, derives the notional from it, then returns the bond, the risk rate and the post-condition that will be committed. Call this before request_warrant: it is free, and it is the only way to learn the cost before committing. The category and the notional are derived from the calldata; they cannot be declared.',
  input: quoteRiskInputSchema,
  output: quoteRiskOutputSchema,
  paid: false,
  readOnly: true,
  async run(client, args) {
    const parsed = parseArgs(quoteRiskInputSchema, args)
    const actionSpec = toActionSpec(parsed.actionSpec)
    const beneficiary = parsed.beneficiary as Address | undefined
    const data = await callGateway(() =>
      client.quote(beneficiary === undefined ? { actionSpec } : { actionSpec, beneficiary }),
    )
    return { kind: 'ok', data }
  },
}

export const requestWarrantTool: WarrantTool<typeof requestWarrantInputSchema, WarrantOpened> = {
  name: 'request_warrant',
  title: 'Open a bonded warrant',
  description:
    'Opens a bonded warrant for the given action and has KeeperHub execute it. Paid: the bond must be funded via x402 before the warrant opens. Returns the warrantId, the executionId and the conditionHash / actionHash commitments. If the post-condition holds, the bond comes back; otherwise it goes to the beneficiary. The bond is derived from the calldata — it is not negotiable.',
  input: requestWarrantInputSchema,
  output: requestWarrantOutputSchema,
  paid: true,
  readOnly: false,
  async run(client, args, ctx) {
    const parsed = parseArgs(requestWarrantInputSchema, args)
    const actionSpec = toActionSpec(parsed.actionSpec)
    const beneficiary = parsed.beneficiary as Address
    const result = await callGateway(() =>
      client.requestWarrant({ actionSpec, beneficiary }, ctx?.payment),
    )
    if (result.status === 'payment-required') {
      return { kind: 'payment-required', paymentRequired: result.paymentRequired }
    }
    return { kind: 'ok', data: result.warrant, settlement: result.settlement }
  },
}

export const getWarrantTool: WarrantTool<typeof getWarrantInputSchema, WarrantView> = {
  name: 'get_warrant',
  title: 'Read a warrant and its verdict',
  description:
    'Returns a warrant, its status and — once it is settled — the full verdict with the checks[] detail: one row per check, including the ones that pass, plus the exact block of evaluation. That is what makes a verdict replayable by a third party rather than taken on trust.',
  input: getWarrantInputSchema,
  output: getWarrantOutputSchema,
  paid: false,
  readOnly: true,
  async run(client, args) {
    const { warrantId } = parseArgs(getWarrantInputSchema, args)
    const view = await callGateway(() => client.getWarrant(warrantId as Hex))
    if (view === null) {
      throw new WarrantError('warrant_not_found', `No warrant under id ${warrantId}.`, {
        field: '$.warrantId',
      })
    }
    return { kind: 'ok', data: view }
  },
}

export const listWarrantsTool: WarrantTool<typeof listWarrantsInputSchema, ListWarrantsResult> = {
  name: 'list_warrants',
  title: 'List an agent warrants and statistics',
  description:
    "Lists an agent's warrants along with their aggregated statistics: number honored, number slashed, total bonded, honor rate. Filterable by status, category and time window. Use it to answer \"what is this agent's track record?\" without reading the chain.",
  input: listWarrantsInputSchema,
  output: listWarrantsOutputSchema,
  paid: false,
  readOnly: true,
  async run(client, args) {
    const query = parseArgs(listWarrantsInputSchema, args)
    const data = await callGateway(() =>
      client.listWarrants({ ...query, agent: query.agent as Address }),
    )
    return { kind: 'ok', data }
  },
}

/** The order is that of docs/09 § 2 — quote, warrant, read, history. */
export const WARRANT_TOOLS = [
  quoteRiskTool,
  requestWarrantTool,
  getWarrantTool,
  listWarrantsTool,
] as const

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyWarrantTool = WarrantTool<z.ZodObject<z.ZodRawShape>, any>

export const WARRANT_TOOL_NAMES = WARRANT_TOOLS.map((t) => t.name) as readonly string[]

export function warrantToolByName(name: string): AnyWarrantTool | undefined {
  return (WARRANT_TOOLS as readonly AnyWarrantTool[]).find((t) => t.name === name)
}

export interface RunToolOptions {
  wallet?: PaymentSigner | undefined
  /** Payment replays allowed. Bounded: an agentic wallet has no human watching. */
  maxPaymentAttempts?: number
}

/**
 * Runs a tool, settling the bond if a wallet is supplied.
 *
 * The payment loop lives here, once, so that neither the HTTP client nor the
 * framework adapters have to reimplement it — that is precisely the kind of
 * duplication that makes five surfaces drift apart.
 */
export async function runTool<T = unknown>(
  client: GatewayClient,
  tool: AnyWarrantTool,
  args: unknown,
  options: RunToolOptions = {},
): Promise<ToolOutcome<T>> {
  const max = options.maxPaymentAttempts ?? 1
  let outcome = (await tool.run(client, args)) as ToolOutcome<T>
  let attempts = 0

  while (outcome.kind === 'payment-required' && options.wallet && attempts < max) {
    attempts += 1
    let payment: PaymentPayload
    try {
      payment = await options.wallet.createPayment(outcome.paymentRequired)
    } catch (err) {
      throw new WarrantError('payment_invalid', `The wallet refused to sign: ${String(err)}`, {
        details: outcome.paymentRequired,
        cause: err,
      })
    }
    outcome = (await tool.run(client, args, { payment })) as ToolOutcome<T>
  }
  return outcome
}

/** Same thing, by tool name. */
export async function runToolByName<T = unknown>(
  client: GatewayClient,
  name: string,
  args: unknown,
  options: RunToolOptions = {},
): Promise<ToolOutcome<T>> {
  const tool = warrantToolByName(name)
  if (!tool) {
    throw new WarrantError('invalid_input', `Unknown tool: ${name}.`, {
      hint: `The available tools are ${WARRANT_TOOL_NAMES.join(', ')}.`,
    })
  }
  return runTool<T>(client, tool, args, options)
}

File: packages/sdk-py/src/warrant_sdk/tools.py
"""The four tools, projected into Python.

Four, not fifteen: the catalogue is deliberately small, because an agent chooses
badly among fifteen tools.

Nothing here defines a schema or a description. Names, titles, descriptions and
JSON Schemas come from ``_generated.py``, which is emitted from
``packages/sdk-ts/src/tools.ts``. What *is* written here is the part that cannot be
generated: how to call the Gateway, and how the x402 payment loop behaves. That
loop lives in one place so that neither the HTTP client nor any framework adapter
has to reimplement it — the kind of duplication that makes five surfaces diverge.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable

from pydantic import BaseModel, ValidationError

from ._generated import MANIFEST_SHA256, TOOL_MANIFEST
from .errors import WarrantError
from .gateway import GatewayClient
from .x402 import PaymentSigner, commitment_terms

__all__ = [
    "MANIFEST_SHA256",
    "ToolOutcome",
    "WarrantTool",
    "WARRANT_TOOLS",
    "WARRANT_TOOL_NAMES",
    "warrant_tool_by_name",
    "run_tool",
    "run_tool_by_name",
]


@dataclass(frozen=True)
class ToolOutcome:
    """The result of a tool call.

    ``payment-required`` is not a domain error: it is step 2 of the x402
    transport. Adapters are free to present it as an error — MCP requires that —
    but the core does not.

    Attributes:
        kind: ``"ok"`` or ``"payment-required"``.
        data: The tool's result, when ``kind == "ok"``.
        settlement: The x402 settlement receipt, when a payment was made.
        payment_required: The x402 challenge, when ``kind == "payment-required"``.
    """

    kind: str
    data: Any = None
    settlement: dict[str, Any] | None = None
    payment_required: dict[str, Any] | None = None


@dataclass(frozen=True)
class WarrantTool:
    """One tool descriptor, framework-agnostic.

    Attributes:
        name: Wire name, e.g. ``quote_risk``.
        title: Short human-readable title.
        description: What the model reads to decide whether to call it.
        paid: ``True`` for ``request_warrant`` only: the bond must be funded.
        read_only: None of these tools mutate onchain state beyond the payment.
        input_model: Pydantic model for the arguments. Drops unknown keys.
        input_schema: JSON Schema draft-7, exactly as MCP's ``tools/list`` publishes it.
        output_schema: Minimal output contract — deliberately permissive.
    """

    name: str
    title: str
    description: str
    paid: bool
    read_only: bool
    input_model: type[BaseModel]
    input_schema: dict[str, Any]
    output_schema: dict[str, Any]

    def parse_args(self, args: Any) -> dict[str, Any]:
        """Validate and **scrub** the arguments.

        The scrubbing is the security-relevant part: the returned mapping contains
        only keys declared by the schema, so a ``category`` slipped into the
        ``actionSpec`` reaches neither the classifier nor the ``actionHash``.

        Raises:
            WarrantError: ``invalid_action_spec`` when the offending field is
                inside the action, ``invalid_input`` otherwise. The distinction is
                not cosmetic — it selects which hint the agent reads, and only the
                first one says that category and notional are derived.
        """
        try:
            model = self.input_model.model_validate(args if args is not None else {})
        except ValidationError as err:
            issues = err.errors()
            paths = ["$." + ".".join(str(p) for p in issue["loc"]) for issue in issues]
            message = "; ".join(
                f"{path}: {issue['msg']}" for path, issue in zip(paths, issues, strict=True)
            )
            first = paths[0] if paths else "$"
            code = "invalid_action_spec" if first.startswith("$.actionSpec") else "invalid_input"
            raise WarrantError(
                code,
                message,
                field=first,
                details=[
                    {"field": path, "message": issue["msg"], "type": issue["type"]}
                    for path, issue in zip(paths, issues, strict=True)
                ],
            ) from err
        # `exclude_none` : un champ optionnel non fourni ne doit pas partir en
        # `null` sur le fil — le Gateway distingue absent et nul.
        return model.model_dump(mode="json", exclude_none=True)

    def run(
        self,
        client: GatewayClient,
        args: Any,
        *,
        payment: dict[str, Any] | None = None,
        warrant_nonce: str | None = None,
    ) -> ToolOutcome:
        """Execute the tool against a Gateway. Does not pay; see :func:`run_tool`."""
        parsed = self.parse_args(args)
        return _RUNNERS[self.name](client, parsed, payment, warrant_nonce)


# ─────────────────────────────────────────────────────────────────────────────
# Les quatre exécutions
# ─────────────────────────────────────────────────────────────────────────────

Runner = Callable[
    [GatewayClient, dict[str, Any], "dict[str, Any] | None", "str | None"], ToolOutcome
]


def _quote_risk(
    client: GatewayClient, parsed: dict[str, Any], _payment: Any, _nonce: Any
) -> ToolOutcome:
    return ToolOutcome(kind="ok", data=client.quote(parsed))


def _request_warrant(
    client: GatewayClient,
    parsed: dict[str, Any],
    payment: dict[str, Any] | None,
    warrant_nonce: str | None,
) -> ToolOutcome:
    result = client.request_warrant(parsed, payment, warrant_nonce=warrant_nonce)
    if result.status == "payment-required":
        return ToolOutcome(kind="payment-required", payment_required=result.payment_required)
    return ToolOutcome(kind="ok", data=result.warrant, settlement=result.settlement)


def _get_warrant(
    client: GatewayClient, parsed: dict[str, Any], _payment: Any, _nonce: Any
) -> ToolOutcome:
    view = client.get_warrant(parsed["warrantId"])
    if view is None:
        raise WarrantError(
            "warrant_not_found",
            f"No warrant under id {parsed['warrantId']}.",
            field="$.warrantId",
        )
    return ToolOutcome(kind="ok", data=view)


def _list_warrants(
    client: GatewayClient, parsed: dict[str, Any], _payment: Any, _nonce: Any
) -> ToolOutcome:
    return ToolOutcome(kind="ok", data=client.list_warrants(parsed))


_RUNNERS: dict[str, Runner] = {
    "quote_risk": _quote_risk,
    "request_warrant": _request_warrant,
    "get_warrant": _get_warrant,
    "list_warrants": _list_warrants,
}


def _build() -> tuple[WarrantTool, ...]:
    """Turn the generated manifest into descriptors.

    Fails at **import** time if the manifest gained a tool this module cannot run.
    A silently missing tool would surface later as a ``KeyError`` inside an agent
    trace, which is the worst place to discover it.
    """
    missing = [spec["name"] for spec in TOOL_MANIFEST if spec["name"] not in _RUNNERS]
    if missing:
        raise RuntimeError(
            f"The tool manifest declares {missing} with no runner in tools.py. "
            "Add one — the manifest is the source, this module follows it."
        )
    return tuple(
        WarrantTool(
            name=spec["name"],
            title=spec["title"],
            description=spec["description"],
            paid=spec["paid"],
            read_only=spec["read_only"],
            input_model=spec["input_model"],
            input_schema=spec["input_schema"],
            output_schema=spec["output_schema"],
        )
        for spec in TOOL_MANIFEST
    )


#: The four tools, in the order of the source: quote, warrant, read, history.
WARRANT_TOOLS: tuple[WarrantTool, ...] = _build()

WARRANT_TOOL_NAMES: tuple[str, ...] = tuple(tool.name for tool in WARRANT_TOOLS)


def warrant_tool_by_name(name: str) -> WarrantTool | None:
    """Look a tool up by its wire name."""
    for tool in WARRANT_TOOLS:
        if tool.name == name:
            return tool
    return None


# ─────────────────────────────────────────────────────────────────────────────
# La boucle de paiement — une seule fois, ici
# ─────────────────────────────────────────────────────────────────────────────


def run_tool(
    client: GatewayClient,
    tool: WarrantTool,
    args: Any,
    *,
    signer: PaymentSigner | None = None,
    max_payment_attempts: int = 1,
) -> ToolOutcome:
    """Run a tool, funding the bond when a signer is available.

    Args:
        signer: Omit it and a 402 comes back as an outcome, untouched. That is the
            right default for a read-only agent: nothing can spend by accident.
        max_payment_attempts: Bounded replays. An unbounded payment loop facing a
            server that answers 402 forever would empty an agent wallet with
            nobody watching.
    """
    outcome = tool.run(client, args)
    attempts = 0

    while outcome.kind == "payment-required" and signer is not None and attempts < max_payment_attempts:
        attempts += 1
        challenge = outcome.payment_required or {}
        terms = commitment_terms(challenge)
        _assert_beneficiary_matches(tool, args, terms)
        try:
            payment = signer.create_payment(challenge)
        except WarrantError:
            raise
        except Exception as err:  # noqa: BLE001 — un signataire tiers peut lever n'importe quoi
            raise WarrantError(
                "payment_invalid",
                f"The signer refused to sign: {err}",
                details=challenge,
            ) from err
        outcome = tool.run(client, args, payment=payment, warrant_nonce=str(terms["nonce"]))

    return outcome


def run_tool_by_name(
    client: GatewayClient,
    name: str,
    args: Any,
    *,
    signer: PaymentSigner | None = None,
    max_payment_attempts: int = 1,
) -> ToolOutcome:
    """Same, by tool name."""
    tool = warrant_tool_by_name(name)
    if tool is None:
        raise WarrantError(
            "invalid_input",
            f"Unknown tool: {name}.",
            hint=f"The available tools are {', '.join(WARRANT_TOOL_NAMES)}.",
        )
    if signer is None:
        signer = getattr(client, "signer", None)
    return run_tool(
        client, tool, args, signer=signer, max_payment_attempts=max_payment_attempts
    )


def _assert_beneficiary_matches(
    tool: WarrantTool, args: Any, terms: dict[str, Any]
) -> None:
    """Refuse to sign terms that name a beneficiary the caller did not ask for.

    The Gateway commits the beneficiary from **its own policy**, not the one in the
    request — and since the EIP-3009 nonce hashes the terms, signing the payment
    signs that beneficiary too. If the two disagree, the agent would be bonding in
    favour of an address it never chose. That is worth a refusal rather than a
    warning: the signature is the only moment where it can still be stopped.
    """
    if not tool.paid:
        return
    requested = (args or {}).get("beneficiary") if isinstance(args, dict) else None
    served = terms.get("beneficiary")
    if requested is None or served is None:
        return
    if str(requested).lower() == str(served).lower():
        return
    raise WarrantError(
        "payment_invalid",
        f"The Gateway commits the bond to beneficiary {served}, but the call asked for "
        f"{requested}. Signing would bond in favour of an address you did not choose.",
        hint=(
            "The beneficiary comes from the Gateway's policy, not from the request. "
            "Either call with the policy's beneficiary, or point WARRANT_BASE_URL at a "
            "Gateway whose policy matches. GET /openapi.json shows the deployment."
        ),
        field="$.beneficiary",
        details={"requested": requested, "committed": served},
    )
